File: 0414-third-maximum-number/solution.java
class Solution {
    public int thirdMax(int[] nums) {
        // int flargest=Integer.MIN_VALUE;
        // int slargest=Integer.MIN_VALUE;
        // int tlargest=Integer.MIN_VALUE;
        // for(int i=0;i<nums.length;i++){
        //     if(nums[i]>flargest){
        //         flargest=nums[i];
        //     }
        // }
        // for(int i=0;i<nums.length;i++){
        //     if(nums[i]>slargest && nums[i]<flargest){
        //         slargest =nums[i];
        //     }
        // }
        // for(int i=0;i<nums.length;i++){
        //     if(nums[i]>tlargest && nums[i]<flargest && nums[i]<slargest ){
        //         tlargest =nums[i];
        //     }
        // }
        // if (tlargest == Integer.MIN_VALUE) {
        //     return (slargest == Integer.MIN_VALUE) ? flargest : slargest;
        // }

        // return tlargest;
        long flargest = Long.MIN_VALUE;
        long slargest = Long.MIN_VALUE;
        long tlargest = Long.MIN_VALUE;

        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > flargest) {
                tlargest = slargest;
                slargest = flargest;
                flargest = nums[i];
            } else if (nums[i] > slargest && nums[i] < flargest) {
                tlargest = slargest;
                slargest = nums[i];
            } else if (nums[i] > tlargest && nums[i] < slargest) {
                tlargest = nums[i];
            }
        }

        // If third largest doesn't exist, return the largest
        return (tlargest == Long.MIN_VALUE) ? (int) flargest : (int) tlargest;
    }
}

File: 0414-third-maximum-number/solution.cpp
class Solution {
public:
    int thirdMax(vector<int>& nums) {
        // long long first = LLONG_MIN;
        // long long second = LLONG_MIN;
        // long long third = LLONG_MIN;
        // for(int x : nums) {
        //     if(x == first || x == second || x == third)
        //         continue;
        //     if(x > first) {
        //         third = second;
        //         second = first;
        //         first = x;
        //     }
        //     else if(x > second) {
        //         third = second;
        //         second = x;
        //     }
        //     else if(x > third) {
        //         third = x;
        //     }
        // }
        // if(third == LLONG_MIN)
        //     return (int)first;
        // return (int)third;
        set<int>st(nums.begin(),nums.end());
        if(st.size()<3){
            return *st.rbegin();
        }
        return *(++++st.rbegin());
    }
};
